### PR TITLE
8136895: Writer not closed with disk full error, file resource leaked

### PR DESCRIPTION
**Pull request by bplb**
In`StreamEncoder::implClose`, move `flushLeftoverChar()` inside the `try` block and the closing of the underlying stream inside the `finally` block.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8136895](https://bugs.openjdk.org/browse/JDK-8136895): Writer not closed with disk full error, file resource leaked (**Bug** - P4)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Vyom Tewari](https://openjdk.org/census#vtewari) (@vyommani - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13503/head:pull/13503` \
`$ git checkout pull/13503`

Update a local copy of the PR: \
`$ git checkout pull/13503` \
`$ git pull https://git.openjdk.org/jdk.git pull/13503/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13503`

View PR using the GUI difftool: \
`$ git pr show -t 13503`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13503.diff">https://git.openjdk.org/jdk/pull/13503.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13503#issuecomment-1512335730)

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/cs/StreamEncoder.java [hunk 1]
```diff
@@ -430,6 +430,10 @@ void implClose() throws IOException {
 
             if (bb.position() > 0)
                 writeBytes();
+        } catch (IOException x) {
+            encoder.reset();
+            throw x;
+        } finally {
             if (ch != null)
                 ch.close();
             else {
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/cs/StreamEncoder.java:439]: If this encoder is for an OutputStreamWriter, and the flush of the left over characters fails, then the finally block will attempt to flush the underlying output stream again. Need to you think about that scenario a bit to decide if the underlying flush should be done for this case or not.

File: test/jdk/java/nio/channels/Channels/Close.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    public static void main(String[] args) throws IOException {
+        boolean failed = false;
+
+        for (String s : new String[] {STR_IOE, STR_MIE}) {
+            System.out.println("string: " + s);
+            ErroringByteChannel channel = new ErroringByteChannel();
+            try (Writer writer = Channels.newWriter
+                    (channel, StandardCharsets.UTF_8.newEncoder(), -1 )) {
+                writer.write(s);
+            } catch(IOException ex) {
```
Comment (turbanoff) [inline test/jdk/java/nio/channels/Channels/Close.java:50]:
```suggestion
            } catch (IOException ex) {
```

File: src/java.base/share/classes/sun/nio/cs/StreamEncoder.java (hunk 1 above)
Comment (bplb) [reply src/java.base/share/classes/sun/nio/cs/StreamEncoder.java:439]: > Need to think about that scenario a bit to decide if the underlying flush should be done for this case or not.

`flushLeftoverChar()` can throw an exception after having invoked `writeBytes()` once or more. This suggests that a flush of the underlying stream might be warranted.

File: test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    public static void main(String[] args) throws IOException {
+        boolean failed = false;
+
+        for (String s : new String[] {STR_IOE, STR_MIE}) {
+            System.out.println("string: " + s);
+            ErroringByteChannel channel = new ErroringByteChannel();
+            try (Writer writer = Channels.newWriter
+                    (channel, StandardCharsets.UTF_8.newEncoder(), -1 )) {
+                writer.write(s);
+            } catch(IOException ex) {
```
Comment (turbanoff) [inline test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java:50]: nit
```suggestion
            } catch (IOException ex) {
```

Comment (bplb) [reply test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java:50]: > nit

This trivial change will be made before integration, thanks.

**Comment by AlanBateman** [comment]
Sill mulling over this one as there is flushing at different levels. There may be left overs chars that need to be flushed, then we have flushing of the underlying output stream for cases where an OutputStreamWriter is built on an output stream that has buffering too.  So I think it looks okay, I'm just wondering if we have missed anything.

**Comment by djelinski** [comment]
If `flushLeftoverChar` throws, `writeBytes` will not be called and any characters left in `bb` will be lost. This should probably be fixed, but given that it's preexisting, it may be deferred to a separate issue.
Encoder.flush won't be called either; the method has a non-trivial implementation for charset ISO-2022-JP, for example. I'm not sure if there's anything we should do about that.

Also, I checked why `out.flush` was added; found [this mail](https://mail.openjdk.org/pipermail/nio-dev/2019-March/005974.html) stating that `StreamEncoderClose` test failed without that line. I checked the test, and it appears to assume that if StreamEncoder catches an exception when closing the underlying stream, it should be closed again. While this might have made sense back when it was implemented, today try-with-resources gives each `Closeable` only one chance to clean up. I still think we should just use `try (ch; out)` and remove the explicit `flush` & `close` calls.

Also, StreamEncoderTest is linked to [JDK-5005426](https://bugs.openjdk.org/browse/JDK-5005426); the bug describes exactly the issue mentioned in my first point above - `flushLeftoverChar` threw an exception and data buffered by `StreamEncoder` was discarded. We could implement a better test for that (e.g. write "Test\ud83c", then close, then check if "Test" was written).

File: test/jdk/java/nio/channels/Channels/Close.java (hunk 2 above)
Comment (bplb) [reply test/jdk/java/nio/channels/Channels/Close.java]: So changed in 4c8ab6d.

File: test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java (hunk 3 above)
Comment (bplb) [reply test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java]: So changed in 4c8ab6d.

File: test/jdk/java/nio/channels/Channels/Close.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.nio.charset.MalformedInputException;
+import java.nio.charset.StandardCharsets;
+
+public class Close {
```
[the remark this answers is not in the document]
Comment (bplb) [reply test/jdk/java/nio/channels/Channels/Close.java]: Addressed in 181a943.

**Comment by bplb** [comment]
> I still think we should just use `try (ch; out)` and remove the explicit `flush` & `close` calls.

Is this what you mean?

```
--- a/src/java.base/share/classes/sun/nio/cs/StreamEncoder.java
+++ b/src/java.base/share/classes/sun/nio/cs/StreamEncoder.java
@@ -436,16 +436,7 @@ public final class StreamEncoder extends Writer {
             ioe = x;
         }
 
-        try {
-            if (ch != null)
-                ch.close();
-            else {
-                try {
-                    out.flush();
-                } finally {
-                    out.close();
-                }
-            }
+        try (ch; out) {
         } catch (IOException e) {
             if (ioe != null)
                 ioe.addSuppressed(e);
```

**Comment by djelinski** [comment]
This is what I had in mind:
```
--- a/src/java.base/share/classes/sun/nio/cs/StreamEncoder.java
+++ b/src/java.base/share/classes/sun/nio/cs/StreamEncoder.java
@@ -414,8 +414,7 @@ public final class StreamEncoder extends Writer {
     }

     void implClose() throws IOException {
-        IOException ioe = null;
-        try {
+        try (ch; out) {
             flushLeftoverChar(null, true);
             for (;;) {
                 CoderResult cr = encoder.flush(bb);
@@ -433,28 +432,7 @@ public final class StreamEncoder extends Writer {
                 writeBytes();
         } catch (IOException x) {
             encoder.reset();
-            ioe = x;
         }
-
-        try {
-            if (ch != null)
-                ch.close();
-            else {
-                try {
-                    out.flush();
-                } finally {
-                    out.close();
-                }
-            }
-        } catch (IOException e) {
-            if (ioe != null)
-                ioe.addSuppressed(e);
-            else
-                ioe = e;
-        }
-
-        if (ioe != null)
-            throw ioe;
     }

     String encodingName() {
```
(yes it still makes `StreamEncoderClose` test fail)

**Comment by bplb** [comment]
> (yes it still makes `StreamEncoderClose` test fail)

`StreamEncoderClose` does not fail with the version of the code currently in the PR (latest commit 4c8ab6df0de5daa71c5864985b83d93b830ba185).

**Comment by bplb** [comment]
continue ...

**Comment by bplb** [comment]
Commit d19659963b8a2a2cea11208459d33f1ca8c1af10 changes `implClose()` along the lines suggested by @djelinski . The `close()` is handled by a `try-with-resources` block. This assumes that any flushing of the channel or output stream would be done by the `close()` of the respective destination. The bytes held in `bb` will not, as mentioned above, be flushed.

This change passes all NIO tests.

**Comment by djelinski** [comment]
Can you remove the explicit calls to `close()`? The underlying stream/channel is now closed twice in the no-exception case.

**Comment by bplb** [comment]
> Can you remove the explicit calls to `close()`? The underlying stream/channel is now closed twice in the no-exception case.

Will do. That was an oversight to leave them in. Thanks for pointing it out.

**Review by djelinski — APPROVED** [review]
LGTM.

**Comment by bplb** [comment]
> LGTM.

Thanks. This change passes all NIO tests as well.

File: test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8136895
+ * @summary Verify channel closed after write error in StreamEncoder::implClose
+ */
```
Comment (vyommani) [inline test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java:27]: Do you mind adding action tag(@run) ?

File: test/jdk/java/io/OutputStreamWriter/CloseWriterOnFailedFlush.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8136895
+ * @summary Verify stream closed after write error in StreamEncoder::implClose
+ */
```
Comment (vyommani) [inline test/jdk/java/io/OutputStreamWriter/CloseWriterOnFailedFlush.java:27]: Do you mind adding action tag(@run) ?

**Review by vyommani — APPROVED** [review]
Code changes looks OK to me, i tested with the Test provided by original submitter and after fix i see underline file descriptors are getting closed.

File: test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java (hunk 5 above)
Comment (bplb) [reply test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java:27]: Why? It is unnecessary. See [jtreg DEFAULTS](https://openjdk.org/jtreg/tag-spec.html#DEFAULTS).

File: test/jdk/java/io/OutputStreamWriter/CloseWriterOnFailedFlush.java (hunk 6 above)
Comment (bplb) [reply test/jdk/java/io/OutputStreamWriter/CloseWriterOnFailedFlush.java:27]: [Ditto](https://openjdk.org/jtreg/tag-spec.html#DEFAULTS).

File: test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java (hunk 5 above)
Comment (vyommani) [reply test/jdk/java/nio/channels/Channels/CloseWriterOnFailedFlush.java:27]: thanks for link, i was not aware of jtreg defaults, looks like i have to update my very old Netbeans plugin it always complains about jtrg tag order etc.

**Pull request closed without merging**